python/dgl/heterograph.py

```diff
@@ -0,0 +1,2094 @@
+        the destination type in the following form:
+
+            {edge_type: edge_specifier}
+
+        where edge_specifier can be either of the following:
+
+        * (source_node_id_tensor, destination_node_id_tensor)
+            * ``source_node_id_tensor`` and ``destination_node_id_tensor`` are
+              IDs within the source and destination node type respectively.
+            * ``edge_type`` is a triplet of
+
+                  (source_node_type_name,
+                   destination_node_type_name,
+                   edge_type_name)
+
+        * scipy.sparse.matrix
+          By default, the rows represent the destination of an edge, and the
+          column represents the source.
+
+    Examples
+    --------
+    Suppose that we want to construct the following heterogeneous graph:
+
+    .. graphviz::
+
+       digraph G {
+           Alice -> Bob [label=follows]
+           Bob -> Carol [label=follows]
+           Alice -> Tetris [label=plays]
+           Bob -> Tetris [label=plays]
+           Bob -> Minecraft [label=plays]
+           Carol -> Minecraft [label=plays]
+           Nintendo -> Tetris [label=develops]
+           Mojang -> Minecraft [label=develops]
+           {rank=source; Alice; Bob; Carol}
+           {rank=sink; Nintendo; Mojang}
+       }
+
+    One can analyze the graph and figure out the metagraph as follows:
+
+    .. graphviz::
+
+       digraph G {
+           User -> User [label=follows]
+           User -> Game [label=plays]
+           Developer -> Game [label=develops]
+       }
+
+    Suppose that one maps the users, games and developers to the following
+    IDs:
+
+        User name   Alice   Bob     Carol
+        User ID     0       1       2
+
+        Game name   Tetris  Minecraft
+        Game ID     0       1
+
+        Developer name  Nintendo    Mojang
+        Developer ID    0           1
+
+    One can construct the graph as follows:
+
+    >>> import networkx as nx
+    >>> metagraph = nx.MultiGraph([
+    ...     ('user', 'user', 'follows'),
+    ...     ('user', 'game', 'plays'),
+    ...     ('developer', 'game', 'develops')])
+    >>> g = DGLBaseHeteroGraph(
+    ...     metagraph=metagraph,
+    ...     number_of_nodes_by_type={'user': 4, 'game': 2, 'developer': 2},
+    ...     edge_connections_by_type={
+    ...         # Alice follows Bob and Bob follows Carol
+    ...         ('user', 'user', 'follows'): ([0, 1], [1, 2]),
+    ...         # Alice and Bob play Tetris and Bob and Carol play Minecraft
+    ...         ('user', 'game', 'plays'): ([0, 1, 1, 2], [0, 0, 1, 1]),
+    ...         # Nintendo develops Tetris and Mojang develops Minecraft
+    ...         ('developer', 'game', 'develops'): ([0, 1], [0, 1])})
+    """
+
+    # pylint: disable=unused-argument
+    def __init__(
+            self,
+            metagraph,
+            number_of_nodes_by_type,
+            edge_connections_by_type):
```

Comment:
How to support converting from other format like networkx, scipy.

Reply:
Added `from_networkx`.  I'm not sure how to handle scipy matrices since it's likely that we'll specify a list of scipy matrices for each edge type (which is handled by `__init__`)